coding/src/analysis.py

```diff
@@ -0,0 +1,423 @@
+        # save arguments
+        self.results_path = results_path
+        # supported matching strategies
+        self.ends_with = ends_with
+        self.matching_strategies = ['starts_with', 'substr']
+        self.matching_strategy = matching_strategy
+        self.normalize_marginal = normalize_marginal
+
+        # load results
+        self.results = None
+        self.load_results()
+        # populate probs
+        self.populate_category_probs()
+        # populate scores
+        self.populate_scores()
+
+    def read_df(self, file_name):
+        '''
+        Reads a file into a pandas dataframe. Supported filetypes: csv, pkl, p, pickle
+        '''
+        if file_name.endswith('.csv'):
+            df = pd.read_csv(file_name)
+        elif file_name.endswith('.pkl') or file_name.endswith('.p') or file_name.endswith('.pickle'):
+            df = pd.read_pickle(file_name)
+        else:
+            raise ValueError('File type not supported. Supported filetypes: csv, pkl')
+        return df
+
+    def read_and_combine_dfs(self, file_names):
+        '''
+        Reads a list of files into a pandas dataframe. Supported filetypes: csv, pkl
+        '''
+        print(file_names)
+        dfs = []
+        for file_name in file_names:
+            dfs.append(self.read_df(file_name))
+        if len(dfs) == 1:
+            df = dfs[0]
+        else:
+            df = pd.concat(dfs, ignore_index=True)
+        return df
+
+    def get_all_files(self, path):
+        '''
+        Searches subdirectories in path and returns a list of all files ending with
+        ends_with.
+        '''
+        matched_files = []
+        for root, dirs, files in os.walk(path):
+            for file in files:
+                # if ends_with is list
+                if isinstance(self.ends_with, list):
+                    for end in self.ends_with:
+                        if file.endswith(end):
+                            matched_files.append(os.path.join(root, file))
+                else:
+                    if file.endswith(self.ends_with):
+                        matched_files.append(os.path.join(root, file))
+        return matched_files
+    
+    def load_results(self):
+        '''
+        Loads all files in results_path into self.results
+        '''
+        files = self.get_all_files(self.results_path)
+        self.results = self.read_and_combine_dfs(files)
+        # dropna in row 'responses'
+        self.results = self.results.dropna(subset=['responses'])
+        # TODO - remove
+        # self.results = self.results.sample(n=1000)
+        # reset index
+        self.results = self.results.reset_index()
+    
+    def get_logprobs(self, response):
+        '''
+        Given an instance of gpt3 response, return the logprobs of the first sampled token.
+        Returns a sorted list of tuples of (token, logprob)
+        '''
+        # if response is tuple, need to extract second element
+        if isinstance(response, tuple):
```

Comment:
This was added because in the first experiment, the column was just 'response', whereas after your indexing change it was (index, response). This way it works in both cases and on old and new experiments.